src/Store.elm

```diff
@@ -35,21 +34,103 @@ type alias Store =
 
 {-| As in, Request
 -}
-type Action
-    = GetPosts
-    | GetUsers
-    | GetImage ImageId
-    | CreatePost PostCreateData
+type alias Action =
+    { run : Store -> ( Store, Cmd Msg )
+    , toastOnSent : Maybe String
+    }
+
+
+type alias ToastMsg =
+    { onFailure : String
+    , onSuccess : Maybe String
+    }
+
+
+base : Action
+base =
+    { run = \store -> ( store, Cmd.none )
+    , toastOnSent = Nothing
+    }
+
+
+getPosts : Action
+getPosts =
+    { base
+        | run =
+            \store ->
+                if shouldSendRequest store.posts then
+                    ( { store | posts = Loading }
+                    , send "Failed to get posts"
+                        (\err s -> { s | posts = Failure err })
+                        API.Post.getAll
+                        GotPosts
+                    )
+
+                else
+                    ( store, Cmd.none )
+    }
+
+
+getUsers : Action
+getUsers =
+    { base
+        | run =
+            \store ->
+                if shouldSendRequest store.users then
+                    ( { store | users = Loading }
+                    , send "Failed to get users"
+                        (\err s -> { s | users = Failure err })
+                        API.User.getAll
+                        GotUsers
+                    )
+
+                else
+                    ( store, Cmd.none )
+    }
+
+
+getImage : ImageId -> Action
+getImage imageId =
+    { base
+        | run =
+            \store ->
+                if shouldSendRequest_ (Dict.get imageId store.images) then
+                    ( { store | images = Dict.insert imageId Loading store.images }
+                    , send ("Failed to get image '" ++ imageId ++ "'")
+                        (\err s ->
+                            { s | images = Dict.insert imageId (Failure err) store.images }
+                        )
+                        (API.Image.get imageId)
+                        GotImage
+                    )
+
+                else
+                    ( store, Cmd.none )
+    }
+
+
+createPost : PostCreateData -> Action
+createPost postCreateData =
+    { run =
+        \store ->
+            ( store
+            , send ("Failed to create post '" ++ postCreateData.title ++ "'")
+                (\_ s -> s)
+                (API.Post.create postCreateData)
+                (CreatedPost ("Created post '" ++ postCreateData.title ++ "'"))
+            )
+    , toastOnSent = Just ("Creating post '" ++ postCreateData.title ++ "'")
+    }
 
 
 {-| As in, Response
 -}
 type Msg
-    = HttpError Action Http.Error -- !
+    = HttpError String Http.Error (Store -> Store) -- !
```

Comment:
Obviously functions in Msg's aren't great and there's a way to get rid of this, can't remember since I submitted this PR ages ago :sweat_smile: